programs/lighthouse/src/types/assert/assertion_result.rs

```diff
@@ -0,0 +1,144 @@
+use super::{EquatableOperator, LogLevel};
+use crate::{error::LighthouseError, types::assert::Operator, validation::SPL_NOOP_ID, Result};
+use borsh::{BorshDeserialize, BorshSerialize};
+use solana_program::{
+    instruction::Instruction, log::sol_log_data, msg, program::invoke, pubkey::Pubkey,
+};
+use std::fmt::Debug;
+
+macro_rules! log_case {
+    ($variant:ident, $actual:ident, $assertion:ident, $operator:ident, $passed:ident) => {
+        msg!(
+            "Result ({}): {:?} {} {:?}",
+            if *$passed { "Passed" } else { "Failed" },
+            $actual,
+            $operator,
+            $assertion,
+        );
+    };
+}
+
+#[derive(BorshDeserialize, BorshSerialize, Debug, Clone)]
+pub enum AssertionResult {
+    U8(Option<u8>, Option<u8>, u8, bool),
+    U16(Option<u16>, Option<u16>, u8, bool),
+    U32(Option<u32>, Option<u32>, u8, bool),
+    U64(Option<u64>, Option<u64>, u8, bool),
+    U128(Option<u128>, Option<u128>, u8, bool),
+    I8(Option<i8>, Option<i8>, u8, bool),
+    I16(Option<i16>, Option<i16>, u8, bool),
+    I32(Option<i32>, Option<i32>, u8, bool),
+    I64(Option<i64>, Option<i64>, u8, bool),
+    I128(Option<i128>, Option<i128>, u8, bool),
+    Pubkey(Option<Pubkey>, Option<Pubkey>, u8, bool),
+    Bytes(Vec<u8>, Vec<u8>, u8, bool),
+    Bool(Option<bool>, Option<bool>, u8, bool),
+}
+
+impl AssertionResult {
+    pub fn log(&self, log_level: LogLevel) -> Result<()> {
+        match log_level {
+            LogLevel::Silent => {}
+            LogLevel::PlaintextMessage => {
+                self.log_msg();
+            }
+            LogLevel::EncodedMessage => {
+                self.log_data()?;
+            }
+            LogLevel::EncodedNoop => {
+                self.log_noop()?;
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn log_data(&self) -> Result<()> {
+        let data = self.try_to_vec().map_err(LighthouseError::serialize_err)?;
+
+        sol_log_data(&[&data]);
+
+        Ok(())
+    }
+
+    pub fn log_noop(&self) -> Result<()> {
+        let data = self.try_to_vec().map_err(LighthouseError::serialize_err)?;
+
+        invoke(
+            &Instruction {
+                program_id: SPL_NOOP_ID,
+                accounts: vec![],
+                data,
+            },
+            &[],
+        )
+    }
+
+    pub fn log_msg(&self) {
+        match self {
+            AssertionResult::U8(actual, assertion, operator, passed) => {
```

Comment:
Can we use a repeat match macro to simplify this?

Reply:
Done